bot/utils/services.py

```diff
@@ -57,5 +113,170 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
+
+    if response is None:
+        return
+
+    if response.status == 200:
+        # Success, return the destination
+        content = await response.json()
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        return _UnfurlReturn(content["destination"], redirects + content["depth"], None, now), response.status
+
+    elif response.status == 400:
+        # Unrecoverable error with the URL used for this unfurling
+        error_message = (await response.json())["error"]
+        log.warning(f"Failed to unfurl URL ({url}) with error message: {error_message}")
+        return _UnfurlReturn(error=error_message), response.status
+
+    elif response.status == 416:
+        # Max depth reached by the worker, try again or exit
+        data = await response.json()
+
+        if continues < max_continues:
+            # Try again (the caller will make a new call with the next URL)
+            log.debug(
+                f"Hit maximum depth while unfurling URL ({url}) after {data['depth']} attempts. "
+                f"Last request was made to ({data['final']}), and next request will be made to {data['next']}. "
+                f"{continues + 1}/{max_continues} failures."
+            )
+
+            return data["next"], data["depth"]
+
+        else:
+            # Give up
+            redirects += data["depth"]
+            log.info(f"Failed to unfurl URL ({url}) after {redirects} redirects.")
+            return _UnfurlReturn(data["next"], redirects, data["error"]), response.status
+
+    elif response.status == 418:
+        # The redirect chain has a broken link, can not continue
+        data = await response.json()
+        redirects += data["depth"]
+
+        log.warning(
+            f"Failed to unfurl URL ({url}) due to a broken link in the chain. "
+            f"Followed {redirects} redirects, and stopped at ({data['final']})."
+        )
+        return _UnfurlReturn(data["final"], redirects, data["error"]), response.status
+
+    else:
+        response.raise_for_status()
+
+
+async def unfurl_url(
+    url: str,
+    *,
+    max_per_attempt: int = MAX_UNFURLS,
+    max_continues: int = 0,
+    calculate_continues: bool = False,
+    use_cache: bool = True,
+) -> Optional[_UnfurlReturn]:
+    """
+    Follow all redirects of a URL, and return the final address.
+
+    Returns an object containing the final URL, number of redirects, error text, and date of last fetch.
+    Returns None if we couldn't resolve the URL for any reason.
+
+    The error in the return is only set if the operation failed.
+    If false, the final destination might not be accurate, and the date will be None.
+
+    `max_per_attempt` is passed directly to the worker. If the worker errors out due to too many redirects,
+    we'll attempt to restart from the last URL `max_continues` times.
+
+    If `calculate_continues` is true, we'll paginate `max_per_attempt` into as many continues as necessary.
```

Comment:
It's still not clear to me what this means. Why does it even need to be configurable?

Reply:
We want to fail quickly if running in filters so the filters aren't slow, but the command may place higher importance on finding a result instead of time.